Bugfix/zenko 1124 fix delete marker handling

jeremyds opened the pull request.

bert-e commented:
# Hello jeremyds,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get
information on this process.





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-1124-fix-delete-marker-handling` into
`development/7.4`
* [w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling](https://github.com/scality/arsenal/compare/development/8.0...w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling) will be merged into `development/8.0` (pull request #560)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @RachedBenMustapha

* @bennettbuchanan

bert-e commented:
# History mismatch



Merge commit #fabbcdb2505513d0f45f66a3ba774055106d7616 on the integration branch
`w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling` is merging a branch which is neither the current
branch `bugfix/ZENKO-1124-fix-delete-marker-handling` nor the development branch
`development/8.0`.

It is likely due to a rebase of the branch `bugfix/ZENKO-1124-fix-delete-marker-handling` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jeremyds commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-1124-fix-delete-marker-handling` into
`development/7.4`
* [w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling](https://github.com/scality/arsenal/compare/development/8.0...w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling) will be merged into `development/8.0` (pull request #561)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @RachedBenMustapha

* @bennettbuchanan

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/Arsenal/commit/8c2ed8129497c7c11f7436db5a949a981f5e2b08) did not succeed in branch bugfix/ZENKO-1124-fix-delete-marker-handling.

bennettbuchanan commented:
@jeremyds It looks like change is affecting other delimiter tests.

Commit:
ZENKO-1124: fix delimiterMaster::filter delete markers handlings.

A specific dataset described and attached in the linked Jira issue shows
a bug in that the current delimiterMaster::filter handling of the delete
marker on a master version and the use of this function into MongoDB
client leads to an infinite loop.

The input dataset, in a versionedis in the form:
- one object with a master version and some others,
- 10k objects with a delete marker on the master version and some other
  versions for each object.
The expected result for this input contains only the first object.

What happens in the delimiterMaster:
- filtering the first object results to adds it to the result and set
  NextMarker to its key. The following versions of this objets are
  skipped and the NextMarker is not modified.
- filtering the others objets results to skip return values.

In the MongoDB client class, after counting 100 skipped values in a row,
the skipscan mechanism is triggered. The problem is that the new range
is computed using NextMarker, which in this case is set to the key of
the 1st object. The listing is done on the same range (well, the first
object is skipped at least).

This commit modifies the skipping value to avoid a client to loop.

bert-e commented:
# History mismatch



Merge commit #8c2ed8129497c7c11f7436db5a949a981f5e2b08 on the integration branch
`w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling` is merging a branch which is neither the current
branch `bugfix/ZENKO-1124-fix-delete-marker-handling` nor the development branch
`development/8.0`.

It is likely due to a rebase of the branch `bugfix/ZENKO-1124-fix-delete-marker-handling` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jeremyds commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-1124-fix-delete-marker-handling` into
`development/7.4`
* [w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling](https://github.com/scality/arsenal/compare/development/8.0...w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling) will be merged into `development/8.0` (pull request #562)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/Arsenal/commit/1cdae2fb7f39f174b3744a08e898aa615efea9a5) did not succeed in branch bugfix/ZENKO-1124-fix-delete-marker-handling.

bert-e commented:
# History mismatch



Merge commit #1cdae2fb7f39f174b3744a08e898aa615efea9a5 on the integration branch
`w/8.0/bugfix/ZENKO-1124-fix-delete-marker-handling` is merging a branch which is neither the current
branch `bugfix/ZENKO-1124-fix-delete-marker-handling` nor the development branch
`development/8.0`.

It is likely due to a rebase of the branch `bugfix/ZENKO-1124-fix-delete-marker-handling` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jeremyds commented:
@bert-e wait

DavidPineauScality commented on lib/algos/list/delimiterMaster.js:
```diff
@@ -116,7 +117,7 @@ class DelimiterMaster extends Delimiter {
 
     skipping() {
-        if (this.NextMarker) {
+        if (this.NextMarker && this.NextMarker !== this.markerParams) {
```
Does this mean that if we haven't updated the NextMarker, `skipping()` will return `SKIP_NONE` ?
I don't see how that improves the situation. Was there a new corner case identified since the last version of the fix ?

vrancurel commented:
wontfix
